src/software/geom/util.cpp

```diff
@@ -1251,6 +1116,151 @@ std::optional<Segment> mergeFullyOverlappingSegments(Segment segment1, Segment s
+    // of the reference Segment
+    std::sort(segments.begin(), segments.end(), [parent_segment](Segment &a, Segment &b) {
+        return (parent_segment.getSegStart() - a.getSegStart()).length() <
+               (parent_segment.getSegStart() - b.getSegStart()).length();
+    });
+
+    // Now we need to find the largest open segment/angle
+    std::vector<Segment> open_segs;
+
+    // The first Angle is between the reference Segment and the first obstacle Segment
+    // After this one, ever open angle is between segment(i).end and
+    // segment(i+1).start
+    open_segs.push_back(
+        Segment(parent_segment.getSegStart(), segments.front().getSegStart()));
+
+    // The 'open' Segment in the space between consecutive 'blocking' Segments
+    for (std::vector<Segment>::const_iterator it = segments.begin();
+         it != segments.end() - 1; it++)
+    {
+        open_segs.push_back(Segment(it->getEnd(), (it + 1)->getSegStart()));
+    }
+
+    // Lastly, the final open angle is between obstacles.end().getEnd() and
+    // reference_segment.getEnd()
+    open_segs.push_back(Segment(segments.back().getEnd(), parent_segment.getEnd()));
+
+    // Remove all zero length open Segments
+    for (std::vector<Segment>::const_iterator it = open_segs.begin();
+         it != open_segs.end();)
+    {
+        if (it->length() < EPS)
+        {
+            open_segs.erase(it);
+        }
+        else
+        {
+            it++;
+        }
+    }
+
+    return open_segs;
+}
+
+
+std::vector<Segment> projectCirclesOntoSegment(Segment segment,
+                                               std::vector<Circle> circles, Point origin)
+{
+    // Loop through all obstacles to create their projected Segment
+    std::vector<Segment> obstacle_segment_projections = {};
+
+    for (Circle circle : circles)
+    {
+        // If the reference is inside an obstacle there is no open direction
+        if (contains(circle, origin))
+        {
+            obstacle_segment_projections.push_back(segment);
+            return obstacle_segment_projections;
+        }
+
+        // Get the tangent rays from the reference point to the obstacle
+        auto [ray1, ray2] = getCircleTangentRaysWithReferenceOrigin(origin, circle);
+
+        // Project the tangent Rays to obtain a 'blocked' segment on the reference
+        // Segment
+        std::optional<Segment> intersect_segment =
+            getIntersectingSegment(ray1, ray2, segment);
+
+        if (intersect_segment.has_value())
+        {
+            obstacle_segment_projections.push_back(intersect_segment.value());
+        }
+    }
+    return obstacle_segment_projections;
+}
+
+std::vector<Segment> combineToParallelSegments(std::vector<Segment> segments,
+                                               Vector direction)
+{
+    std::vector<Segment> projected_segments = {};
+    // The projection of the Segment without including the original Segment location
+    Vector raw_projection = Vector(0, 0);
```

Comment:
Why is this declared outside the `for` loop? I think we should be be able to combine it with the `raw_projection = ....` line below?

Reply:
I originally did that to prevent it from getting re-initialized every loop.

I've moved it into the loop now